# Biological Plots

* [Setup](#setup)
* [Genotype Distribution](#genotype-distribution)
* [Callability](#callability)
* [WGS Statistics](#wgs-statistics)
* [Allele Frequency Distribution](#allele-frequency)
* [BRCA1 Mutation Spectrum](#brca1-mutation-spectrum)
* [Variant Distribution](#variant-distribution)
* [Pathogenic Variants on ACMG Genes](#acgm-genes)
* [Allele Frequency Distributions vs 1000 Genomes](#af-1kg)

## Setup
```{r echo=FALSE, eval=FALSE}
require(scales)
require(reshape2)
require(dplyr)
require(plyr)

# Set Working Directory
setwd("/Users/gmcinnes/src/mvp_aaa_codelabs/misc")

# Set the Google Cloud Platform project id under which these queries will run.
project <- "gbsc-gcp-project-mvp"

# Set up for BigQuery access.
source("../rHelpers/setup.R")
source("../rHelpers/multiplot.R")

# Read in sample information
sampleData <- read.csv("../data/patient_info.csv")
sampleInfo <- select(sampleData, call_call_set_name=Catalog.ID, gender=Gender)

chromosomeLengths <- read.table("http://genome.ucsc.edu/goldenpath/help/hg19.chrom.sizes")
names(chromosomeLengths) = c("reference_name", "length")
```

Plot theme
```{r}
plot_theme = theme_minimal(base_size = 14, base_family = "Helvetica") + 
  theme(axis.line = element_line(colour = "black"),
        panel.grid = element_blank())
```

```{r}
boxPlotTheme = theme_minimal(base_size=14, base_family = "Helvetica") +
  theme(panel.grid = element_blank())
```

Tables for biological queries
```{r}
queryReplacements <- list("_THE_TABLE_"="va_aaa_pilot_data.genome_calls_full_qc",
                          "_THE_EXPANDED_TABLE_"="va_aaa_pilot_data.multi_sample_variants_full_qc",
                          "_BLACKLISTED_TABLE_"="resources.blacklisted_positions",
                          "_GENOTYPING_TABLE_"="va_aaa_pilot_data.genotyping_data",
                          "_ANNOVAR_TABLE_"="resources.annovar_hg19_aaa",
                          "_CHR_LENGTHS_"="stanford.edu:gbsc-stanford-google:resources.hg19_Assembly_BinaRuns",
                          "_PATIENT_INFO_"="va_aaa_pilot_data.patient_info",
                          "_ACMG_GENES_"="stanford.edu:gbsc-stanford-google:resources.56ACMGgenes_Tx")
```



## Genotype Distribution
#### Genotype Counts per Chromosome
```{r message=FALSE, warning=FALSE, comment=NA}
genotypeCountResult <- DisplayAndDispatchQuery("../sql/genotype-counts.sql",
                                               project=project,
                                               replacements=queryReplacements)
```

Setup
```{r}
genotypeCountResult <- genotypeCountResult[complete.cases(genotypeCountResult),]
genotypeCountResult$Cnt = genotypeCountResult$Cnt/1000000
```

```{r genotype-counts-publication, fig.align="center", fig.width=10, message=FALSE, comment=NA}
counts = ggplot(genotypeCountResult) +
  geom_bar(aes(x=Genotype, y=Cnt), stat="identity") +
  xlab("Genotypes") + 
  ylab("SNV Count (millions)") + 
  scale_y_continuous(labels=comma, expand = c(0, 0)) + 
  plot_theme
counts
```

#### Variant Counts By Chromosome
```{r message=FALSE, warning=FALSE, comment=NA}
variantCountResult <- DisplayAndDispatchQuery("../sql/variants-by-chromosome.sql",
                                              project=project,
                                              replacements=queryReplacements)
```

Setup
```{r}
variantCountResult <- join(variantCountResult, chromosomeLengths, by = "reference_name")
variantCountResult$scaled_count <- variantCountResult$Cnt / variantCountResult$length

chromosomes <- c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9", "chr10", "chr11", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17", "chr18", "chr19", "chr20", "chr21", "chr22", "chrX", "chrY", "chrM")
snps <- variantCountResult[grep("SNV", variantCountResult$VAR_type), ]
snps$reference_name <- factor(snps$reference_name, levels=chromosomes)
snps <- snps[complete.cases(snps),]
snps$Cnt <- snps$Cnt/1000000
indels <- variantCountResult[grep("INDEL", variantCountResult$VAR_type), ]
indels$reference_name <- factor(indels$reference_name, levels=chromosomes)
indels <- indels[complete.cases(indels),]



```

```{r snv-counts-publication, fig.align="center", fig.width=10, message=FALSE, comment=NA}
snvs = ggplot(data=snps, aes(y=Cnt, x=reference_name)) + 
  geom_point(size=4) + 
  ylab("SNV Count (millions)") +
  xlab("Chromosome") +
  scale_y_continuous(label=comma) +
  plot_theme +
  theme(axis.text.x = element_text(angle=90, vjust=1)) 
snvs
```

```{r indel-counts-publication, fig.align="center", fig.width=10, message=FALSE, comment=NA}
indels = ggplot(data=indels, aes(y=Cnt, x=reference_name)) + 
  geom_point(size=4) + 
  ylab("Indel Count") +
  xlab("Chromosome") +
  scale_y_continuous(label=comma) +
  plot_theme +
  theme(axis.text.x = element_text(angle=90, vjust=1)) 
indels
```

#### Saturation Rate
```{r message=FALSE, warning=FALSE, comment=NA}
# Need to redo queries for this
genomeCount = c(1,2,3,4,5,10,50,100,200,300,400,478)
snvCount = c(3590360,4847512,5627244,6158953,6616457,
              8014799,11841547,14387937,18693833,21567571,23638061,25890797)
saturationRate = data_frame(genomeCount, snvCount)
saturationRate$snvCount = saturationRate$snvCount/1000000
```


```{r saturation-publication, fig.align="center", fig.width=10, message=FALSE, comment=NA}
saturation = ggplot(saturationRate) +
  geom_point(aes(x=genomeCount, y=snvCount), size=4) +
  xlab("Number of Genomes") +
  ylab("Unique SNVs (millions)") +
  scale_y_continuous(label=comma) +
  plot_theme
saturation
```

#### Multiplot
```{r variant-multiplot-publication, fig.align="center", fig.width=14, message=FALSE, comment=NA}
multiplot(snvs,indels, counts, saturation, cols=2)
```


## Callability
```{r message=FALSE, warning=FALSE, comment=NA}
callability <- DisplayAndDispatchQuery("../sql/callability.sql",
                                              project=project,
                                              replacements=queryReplacements)
```

Setup
```{r}
sample = 'LP6005038-DNA_A01'
selection = callability[callability$call_call_set_name == sample,]
selection$reference_name <- factor(selection$reference_name, levels=chromosomes)
selection <- selection[complete.cases(selection),]
selection$remainder <- selection$contig_len - selection$pos_no_point_info 

faceted = melt(selection, id.vars=c("reference_name"),measure.vars=c("prop_w_point_ino","pos_no_point_info","remainder"))
faceted$color = faceted$variable
faceted[faceted$variable=="remainder",]$variable <- 'pos_no_point_info'
faceted$variable <- factor(faceted$variable)
levels(faceted$variable) <- c("Proportion of positions called", "Number of missing positions")
levels(faceted$color) <- c("Proportion of positions called", "Number of missing positions", "Chromosome length")
```

```{r callability, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(faceted) + 
  geom_bar(aes(reference_name,value, fill=color), stat="identity") +
  facet_grid(variable ~ ., scales="free_y") +
  scale_y_continuous(labels=comma, expand=c(0,0)) +
  plot_theme +
  xlab("Chromosome") +
  scale_fill_brewer(palette=1) +
  theme(axis.text.x = element_text(angle=90, vjust=0.5, hjust=1),
    axis.title.y=element_blank(),
    legend.title=element_blank())
```


## WGS Statistics

#### Ti/Tv for SNVs
```{r message=FALSE, warning=FALSE, comment=NA}
titvBySample <- DisplayAndDispatchQuery("../sql/titv-by-sample.sql",
                                              project=project,
                                              replacements=queryReplacements)
```

```{r titv-by-sample, fig.align="center", fig.width=10, message=FALSE, comment=NA}
titv = ggplot(titvBySample) +
  geom_boxplot(aes(x="SNV", y=titv)) +
  ylab("Ti/Tv") +
  ggtitle("Ti/Tv for SNVs in each genome") +
  boxPlotTheme +
  theme(axis.text.x=element_blank(),
        axis.title.x=element_blank())
titv
```

#### Het/Hom for SNVs
```{r message=FALSE, warning=FALSE, comment=NA}
hetHomSnv <- DisplayAndDispatchQuery("../sql/hethom-snv.sql",
                                              project=project,
                                              replacements=queryReplacements)
```

```{r hethom-snv, fig.align="center", fig.width=10, message=FALSE, comment=NA}
hethom= ggplot(hetHomSnv) +
  geom_boxplot(aes(x="SNV", y=Het_Hom_ratio)) +
  ylab("Het/Hom") +
  ggtitle("Het/Hom for SNVs in each genome") +
  boxPlotTheme + 
  theme(axis.text.x=element_blank(),
        axis.title.x=element_blank())
hethom
```

#### SNVs in dbSNP
```{r message=FALSE, warning=FALSE, comment=NA}
dbSNPcount <- DisplayAndDispatchQuery("../sql/dbsnp-count.sql",
                                              project=project,
                                              replacements=queryReplacements)
dbSNPcount$proportion = dbSNPcount$num_VAR_dbSNP/dbSNPcount$num_VAR
```

```{r dbsnp-snv, fig.align="center", fig.width=10, message=FALSE, comment=NA}
dbSNP = ggplot(dbSNPcount) +
  geom_boxplot(aes(x=VAR_type, y=proportion)) +
  ylab("Proportion of Variants in dbSNP") +
  ggtitle("Proportion of variants in dbSNP by variant type") +
  scale_y_continuous(label=comma) +
  boxPlotTheme +
  theme(axis.title.x=element_blank())
dbSNP
```

#### Private SNV Counts
```{r message=FALSE, warning=FALSE, comment=NA}
privateSNVs <- DisplayAndDispatchQuery("../sql/private-snv-counts.sql",
                                              project=project,
                                              replacements=queryReplacements)
```

```{r private-snv, fig.align="center", fig.width=10, message=FALSE, comment=NA}
privateVariants = ggplot(privateSNVs) +
  geom_boxplot(aes(x="SNV", y=private_SNVs_count)) +
  ylab("Private Variants") +
  ggtitle("Number of private variants per genome") +
  scale_y_continuous(label=comma) +
  boxPlotTheme + 
  theme(axis.text.x=element_blank(),
        axis.title.x=element_blank())
privateVariants 
```

#### Multiplot
```{r boxplots-multiplot-publication, fig.align="center", fig.width=14, message=FALSE, comment=NA}
multiplot(titv, dbSNP, hethom, privateVariants, cols=2)
```


## Allele Frequency Distribution
```{r message=FALSE, warning=FALSE, comment=NA}
rarity <- DisplayAndDispatchQuery("../sql/variant-rarity.sql",
                                              project=project,
                                              replacements=queryReplacements)

rarity$total = sapply(1:nrow(rarity), function(x){
  chr = rarity[x,]$reference_name
  variantCount = sum(rarity[rarity$reference_name == chr,]$count)
  variantCount
})
rarity$proportion = rarity$count/rarity$total
rarity$reference_name <- factor(rarity$reference_name, levels=chromosomes)
rarity$rarity <- factor(rarity$rarity, levels = c("very_rare", "rare", "relatively_common", "common"), ordered=TRUE)
```

```{r variant-rarity, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(rarity) +
  geom_bar(aes(factor(reference_name), proportion, fill=rarity, order=rarity), stat='identity', position='stack') +
  xlab("Chromosome") +
  ylab("Proportion") +
  ggtitle("Allele frequency distribution") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_fill_brewer(palette=1, breaks=c("common","relatively_common","rare", "very_rare"),
                    labels=c("Common", "Relatively common", "Rare", "Very rare")) +
  boxPlotTheme + 
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),
        legend.title=element_blank())
```


## P53 Mutation Spectrum
```{r}
require(NMF)
require(RColorBrewer)
```

```{r message=FALSE, warning=FALSE, comment=NA}
mutationSpectrum <- DisplayAndDispatchQuery("../sql/mutation-spectrum-p53.sql",
                                              project=project,
                                              replacements=queryReplacements,
                                              max=Inf)
```

Get Patient Info
```{r message=FALSE, warning=FALSE, comment=NA}
patientInfo <- DisplayAndDispatchQuery("../sql/patient-info.sql",
                                              project=project,
                                              replacements=queryReplacements,
                                              max=Inf)
```

Convert mutation information to matrix 
```{r}
spectrum <- dcast(mutationSpectrum, call_call_set_name ~ window_start, value.var = "num_variants_in_window", na.rm=TRUE)
spectrum[is.na(spectrum)] <- 0

spectrumMatrix = as.matrix(spectrum[,!colnames(spectrum) %in% c("call_call_set_name")])
rownames(spectrumMatrix) = spectrum$call_call_set_name

sampleIds = data.frame(rownames(spectrumMatrix))
names(sampleIds) = 'sample_id'

annotations = merge(x = sampleIds, y = patientInfo, by = "sample_id", all.x = TRUE)
annotations = annotations[order(annotations$COHORT),]
spectrum = spectrum[annotations$sample_id,]

```

Set up for plot
```{r}
colors = c("blue", "red")
names(colors) = c("CASE", "CONTROL")
annotationColors = list(Cohort=colors)

cohort = data.frame(annotations$COHORT)
names(cohort) = "Cohort"
```

```{r mutation-spectrum, fig.align="center", fig.width=10, message=FALSE, comment=NA}
aheatmap(t(spectrumMatrix), Rowv=NA, Colv=NA, color="Reds:50",labCol="",main="P53 mutation spectrum in 1kb windows", fontsize=14)

aheatmap(t(spectrumMatrix), Rowv=NA, Colv=FALSE, color="Reds:50",labCol="",main="P53 mutation spectrum in 1kb windows", fontsize=14)
```


## Variant Distribution
```{r message=FALSE, warning=FALSE, comment=NA}
regionalCounts <- DisplayAndDispatchQuery("../sql/regional-variant-count.sql",
                                              project=project,
                                              replacements=queryReplacements,
                                              max=Inf)
regionalCounts$Chromosome <- factor(regionalCounts$Chromosome, levels=chromosomes)
regionalCounts$scaled <- regionalCounts$cnt/regionalCounts$chromosome_length
regionalCounts <- regionalCounts[complete.cases(regionalCounts),]
```

#### Exonic
```{r exonic, fig.align="center", fig.width=10, message=FALSE, comment=NA}
exonic = ggplot(regionalCounts) +
  geom_boxplot(data=subset(regionalCounts,region=="exonic"), aes(Chromosome,scaled)) +
  ggtitle("Variants in Exonic Regions") +
  boxPlotTheme + 
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank(),
        axis.title=element_blank())
exonic
```

#### Intronic
```{r intronic, fig.align="center", fig.width=10, message=FALSE, comment=NA}
intronic = ggplot(regionalCounts) +
  geom_boxplot(data=subset(regionalCounts,region=="intronic"), aes(Chromosome,scaled)) +
  ggtitle("Variants in Intronic Regions") +
  boxPlotTheme + 
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank(),
        axis.title=element_blank())
intronic
```

#### UTR3
```{r utr3, fig.align="center", fig.width=10, message=FALSE, comment=NA}
utr3 = ggplot(regionalCounts) +
  geom_boxplot(data=subset(regionalCounts,region=="UTR3"), aes(Chromosome,scaled)) +
  ggtitle("Variants in UTR3 Regions") +
  boxPlotTheme + 
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank(),
        axis.title=element_blank())
utr3
```

#### UTR5
```{r utr5, fig.align="center", fig.width=10, message=FALSE, comment=NA}
utr5 = ggplot(regionalCounts) +
  geom_boxplot(data=subset(regionalCounts,region=="UTR5"), aes(Chromosome,scaled)) +
  ggtitle("Variants in UTR5 Regions") +
  boxPlotTheme + 
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank(),
        axis.title=element_blank())
utr5
```

#### Intergenic
```{r intergenic, fig.align="center", fig.width=10, message=FALSE, comment=NA}
intergenic = ggplot(regionalCounts) +
  geom_boxplot(data=subset(regionalCounts,region=="intergenic"), aes(Chromosome,scaled)) +
  ggtitle("Variants in Intergenic Regions") +
  boxPlotTheme + 
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank(),
        axis.title=element_blank())
intergenic
```

#### Splicing
```{r splicing, fig.align="center", fig.width=10, message=FALSE, comment=NA}
splicing = ggplot(regionalCounts) +
  geom_boxplot(data=subset(regionalCounts,region=="splicing"), aes(Chromosome,scaled)) +
  ggtitle("Variants in Splicing Regions") +
  boxPlotTheme + 
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank(),
        axis.title=element_blank())
splicing
```

#### Multiplot
#### Normalized variant counts by genomic region
```{r regional-counts, fig.align="center", fig.width=14, message=FALSE, comment=NA}
multiplot(exonic,utr3, intergenic, intronic, utr5, splicing, cols=2)
```


## Pathogenic Variants on ACMG Genes
```{r message=FALSE, warning=FALSE, comment=NA}
acmgVariants <- DisplayAndDispatchQuery("../sql/acmg-pathogenic-variants-counts.sql",
                                              project=project,
                                              replacements=queryReplacements)
```

#### Setup
```{r}
acmgExpanded <- dcast(acmgVariants, Sample_id ~ Gene, value.var = "count", na.rm=TRUE)
sampleIds = data.frame(acmgExpanded$Sample_id)
names(sampleIds) = "sample_id"
annotations = merge(x = sampleIds, y = patientInfo, by = "sample_id", all.x = TRUE)
annotations = annotations[order(annotations$COHORT),]
acmgExpanded = acmgExpanded[annotations$sample_id,]
acmgExpanded[is.na(acmgExpanded)] <- 0
acmgMatrix = as.matrix(acmgExpanded[,!colnames(acmgExpanded) %in% c("Sample_id")])
rownames(acmgMatrix) = acmgExpanded$Sample_id
```

Setup for plot
```{r}
sampleIds = data.frame(rownames(acmgMatrix))
names(sampleIds) = 'sample_id'

colors = c("blue", "red")
names(colors) = c("CASE", "CONTROL")
annotationColors = list(Cohort=colors)

cohort = data.frame(annotations$COHORT)
names(cohort) = "Cohort"
```

```{r acmg-variants, fig.align="center", fig.width=10, message=FALSE, comment=NA}
aheatmap(t(acmgMatrix), Rowv=FALSE, Colv=FALSE, color="Reds:2",labCol="",main="Pathogenic variants within ACMG genes", fontsize=14)
```



## Allele Frequency Distributions vs 1000 Genomes
```{r message=FALSE, warning=FALSE, comment=NA}
alleleFrequencies <- DisplayAndDispatchQuery("../sql/1kg-AAA-variant-overlap.sql",
                                              project=project,
                                              replacements=queryReplacements)
```

Setup
```{r}
frequencies = melt(alleleFrequencies, id=c("AGREEMENT","COUNT"))
frequencies = frequencies[complete.cases(frequencies),]
frequencies = frequencies[!(frequencies$AGREEMENT != "DISAGREEMENT" & frequencies$variable == "AAA"),]
frequencies$variable <- factor(frequencies$variable, levels=c("AAA","K1G","BOTH"))
frequencies[frequencies$AGREEMENT != "DISAGREEMENT",]$variable = "BOTH"
names(frequencies) = c("agreement", "count", "population", "frequency")
frequencies$frequency <- factor(frequencies$frequency, levels=c("HIGH","MODERATE","LOW"))
frequencies$normalized = sapply(1:nrow(frequencies), function(x){
  freq = frequencies[x,]$frequency
  freqTotal = sum(frequencies[frequencies$frequency == freq,]$count)
  normalized = frequencies[x,]$count/freqTotal
  normalized
})
```

```{r allele-frequencies-1kg, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(frequencies) +
  geom_bar(aes(frequency, normalized, fill=population, order=population), stat="identity") +
  xlab("Allele frequency") +
  ylab("Proportion of total count") +
  ggtitle("Allele frequency overlap between\ndeeply sequenced genomes and the\nEuropean subpopulation in 1000 Genomes") +
  plot_theme + 
  scale_y_continuous(expand = c(0, 0)) +
  scale_fill_brewer(palette=1, breaks=c("AAA","K1G","BOTH"),
                    labels=c("This study", "1000 Genomes specific", "Overlapping alleles")) +
  theme(legend.title=element_blank()) +
  guides(fill = guide_legend(reverse = TRUE)) +
  scale_x_discrete(breaks = c("HIGH", "MODERATE", "LOW"), labels=c(">0.05", "0.005-0.05","0.001-0.005"))
```

## Query run times
```{r message=FALSE, warning=FALSE, comment=NA}
runtimes = read.csv('../data/query_run_times.csv')
```

Setup
```{r}
dataTraversed = melt(runtimes, id.vars = "Query", measure.vars=c("DataTraversed5Genomes","DataTraversed460Genomes"))
times = melt(runtimes, id.vars = "Query", measure.vars=c("AverageRunTime5Genomes","AverageRunTime460Genomes"))
times$plot="Average Run Time (seconds)"
dataTraversed$plot="Data Processed (gigabytes)"
queryInfo = rbind(times,dataTraversed)
levels(queryInfo$variable) <- c(levels(queryInfo$variable), "5 Genomes", "460 Genomes")
queryInfo[queryInfo == "AverageRunTime5Genomes"] <- "5 Genomes"
queryInfo[queryInfo == "AverageRunTime460Genomes"] <- "460 Genomes"
queryInfo[queryInfo == "DataTraversed5Genomes"] <- "5 Genomes"
queryInfo[queryInfo == "DataTraversed460Genomes"] <- "460 Genomes"
levels(queryInfo$variable) <- c("5 Genomes", "460 Genomes")
queryInfo$variable <- factor(queryInfo$variable)
```
cbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
```{r run-time, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(queryInfo) +
  geom_bar(aes(Query, value, fill=variable), stat="identity", position="dodge") +
  facet_grid(plot ~ ., scales="free_y") +
  plot_theme +
  theme(axis.text.x = element_text(angle=90, vjust=0.5, hjust=1),
  axis.title.y=element_blank(),
  legend.title=element_blank()) +
  scale_fill_hue(l=40) +
  ggtitle("Query processing information in BigQuery") +
  scale_y_continuous(labels=comma, expand = c(0, 0))  
```

## Warfarin Dosage
```{r message=FALSE, warning=FALSE, comment=NA}
warfarin <- DisplayAndDispatchQuery("../sql/warfarin-dosage.sql",
                                              project=project,
                                              replacements=queryReplacements)
```

```{r warfarin, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(warfarin) +
  geom_bar(aes('dosage',count, fill=dosage), width=1, stat="identity") +
  coord_polar(theta='y') +
  boxPlotTheme +
  ggtitle("Recommended daily warfarin dosage") +
  theme(axis.ticks=element_blank(),
            axis.text=element_blank(),
            axis.title=element_blank(),
            panel.border=element_blank(),
            plot.title = element_text(vjust=-2)) 
```
