# Publication Plots

## Setup
```{r}
require(scales)
require(reshape2)
require(dplyr)
require(plyr)

# Set Working Directory
setwd("/Users/gmcinnes/src/mvp_aaa_codelabs/qc")

# Set the Google Cloud Platform project id under which these queries will run.
project <- "gbsc-gcp-project-mvp"

# Set up for BigQuery access.
source("./rHelpers/setup.R")

# Define Query Substitutions
queryReplacements <- list("_THE_TABLE_"="va_aaa_pilot_data.all_genomes_gvcfs",
                          "_THE_EXPANDED_TABLE_"="va_aaa_pilot_data.all_genomes_expanded_vcfs_java2",
                          "_BLACKLISTED_TABLE_"="resources.blacklisted_positions",
                          "_GENOTYPING_TABLE_"="va_aaa_pilot_data.genotyping_data")

# Read in sample information
sampleData <- read.csv("./data/patient_info.csv")
sampleInfo <- select(sampleData, call_call_set_name=Catalog.ID, gender=Gender)

chromosomeLengths <- read.table("http://genome.ucsc.edu/goldenpath/help/hg19.chrom.sizes")
names(chromosomeLengths) = c("reference_name", "length")
```


## Biological Query Plots
#### Genotype Counts
```{r message=FALSE, warning=FALSE, comment=NA}
genotypeCountResult <- DisplayAndDispatchQuery("./sql/genotype-counts.sql",
                                             project=project,
                                             replacements=queryReplacements)
genotypeCountResult <- genotypeCountResult[complete.cases(genotypeCountResult),]
```

```{r genotype-counts-publication, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(genotypeCountResult) +
  geom_bar(aes(x=Genotype, y=Cnt), stat="identity") +
  xlab("Genotypes") + 
  ylab("SNV Count") + 
  ggtitle("SNV Counts for Each Genotype") +
  scale_y_continuous(labels=comma, expand = c(0, 0)) +
  theme(axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        axis.text=element_text(size=24),
        axis.title=element_text(size=28),
        title=element_text(size=32,face="bold"),
        legend.text=element_text(size=24)) 
```

#### Variant Counts By Chromosome
```{r message=FALSE, warning=FALSE, comment=NA}
variantCountResult <- DisplayAndDispatchQuery("./sql/variants-by-chromosome.sql",
                                             project=project,
                                             replacements=queryReplacements)

variantCountResult <- join(variantCountResult, chromosomeLengths, by = "reference_name")
variantCountResult$scaled_count <- variantCountResult$Cnt / variantCountResult$length

chromosomes <- c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9", "chr10", "chr11", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17", "chr18", "chr19", "chr20", "chr21", "chr22", "chrX", "chrY", "chrM")
snps <- variantCountResult[grep("SNV", variantCountResult$VAR_type), ]
snps$reference_name <- factor(snps$reference_name, levels=chromosomes)
snps <- snps[complete.cases(snps),]

indels <- variantCountResult[grep("INDEL", variantCountResult$VAR_type), ]
indels$reference_name <- factor(indels$reference_name, levels=chromosomes)
indels <- indels[complete.cases(indels),]



```

```{r snv-counts-publication, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(data=snps, aes(y=scaled_count, x=reference_name)) + 
  geom_point() + 
  ylab("Scaled SNV Count") +
  xlab("Chromosome") +
  scale_y_continuous(label=comma) +
  theme(text = element_text(size=12), 
        axis.text.x = element_text(angle=90, vjust=1),
        axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        axis.text=element_text(size=20),
        axis.title=element_text(size=28),
        title=element_text(size=32,face="bold"),
        legend.text=element_text(size=24)) 
```

```{r indel-counts-publication, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(data=indels, aes(y=scaled_count, x=reference_name)) + 
  geom_point() + 
  ylab("Scaled Indel Count") +
  xlab("chromosome") +
  scale_y_continuous(label=comma) +
  theme(text = element_text(size=12), 
        axis.text.x = element_text(angle=90, vjust=1),
        axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        axis.text=element_text(size=20),
        axis.title=element_text(size=28),
        title=element_text(size=32,face="bold"),
        legend.text=element_text(size=24)) 
```

#### Saturation Rate
```{r message=FALSE, warning=FALSE, comment=NA}
# Need to redo queries for this
genome_count = c(1,2,3,4,5,10,50,100,200,300,400,478)
snv_count = c(3590360,4847512,5627244,6158953,6616457,
              8014799,11841547,14387937,18693833,21567571,23638061,25890797)
saturation_rate = data_frame(genome_count, snv_count)

```


```{r saturation-publication, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(saturation_rate) +
  geom_point(aes(x=genome_count, y=snv_count), size=4) +
  xlab("Number of Genomes") +
  ylab("Unique SNVs") +
  scale_y_continuous(label=comma) +
  ggtitle("Saturation Rate") +
  theme(axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        axis.text=element_text(size=24),
        axis.title=element_text(size=28),
        title=element_text(size=32,face="bold"),
        legend.text=element_text(size=24)) 
```

## QC Plots
#### Genotyping Concordance
```{r message=FALSE, warning=FALSE, comment=NA}
concordanceResult <- DisplayAndDispatchQuery("./sql/genotyping-concordance.sql",
                                             project=project,
                                             replacements=queryReplacements)
plate = substr(concordanceResult$sample_id, 1, 9)
concordanceResult = cbind(concordanceResult, plate)
```

```{r concordance-publication, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(concordanceResult) +
  geom_point(aes(x=sample_id, y=concordance, color=plate), size=4) +
  xlab("Sample") +
  ylab("Concordance") +
  ggtitle("Concordance with Genotyping Data") +
  scale_colour_brewer(name="Sample Prep Plate", palette="Set1") +
  theme(axis.text.x=element_blank(),
        axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        axis.text=element_text(size=24),
        axis.title=element_text(size=28),
        axis.ticks = element_blank(),
        axis.text.x = element_blank(),
        title=element_text(size=32,face="bold"),
        legend.text=element_text(size=24)) 
```

####Batch Effect PCA
```{r message=FALSE, warning=FALSE, comment=NA}
pcaFile = '/Users/gmcinnes/data/pca-all-genomes-all-references-no-tissue.tsv'
pcaResult = read.table(pcaFile)
names(pcaResult) = c('sample_id','pc1', 'pc2', 'something')
plate = substr(pcaResult$sample_id, 1, 9)
pcaResult = cbind(pcaResult, plate)
```

```{r pca-batch-effect-publication, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(pcaResult, aes(pc1, pc2, color=plate)) + 
  geom_point(size=7) +
  ggtitle("Principal Component Analysis") +
  xlab("Principal Component 1") + 
  ylab("Principal Component 2") +
  scale_colour_brewer(name="Sample Prep Plate", palette="Set1") +
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        axis.title=element_text(size=28),
        title=element_text(size=32,face="bold"),
        legend.text=element_text(size=24)) 
```

#### Ti/Tv By Depth
```{r message=FALSE, warning=FALSE, comment=NA}
query <- "./sql/ti-tv-by-depth.sql"
titv <- DisplayAndDispatchQuery(query,
                                project=project,
                                replacements=c(queryReplacements))
```

```{r titv-by-depth-publication, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(titv, aes(x=average_depth, y=titv_ratio, color=call_call_set_name)) + 
  geom_point(size=3) +
  ggtitle("Ti/Tv Ratio By Depth") +
  xlab("Coverage Depth") + 
  ylab("Ti/Tv") +
  theme_bw() +
  scale_x_continuous(expand = c(0, 0)) +
  theme(axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        legend.position = "none",
        axis.text=element_text(size=24),
        axis.title=element_text(size=28),
        title=element_text(size=32,face="bold"))
```

#### Sex Inference
```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("./sql/gender-check.sql",
                                  project=project,
                                  replacements=queryReplacements)

joinedResult <- inner_join(result, sampleInfo)
```

```{r sex-inference-publication, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(joinedResult) +
  geom_point(aes(x=call_call_set_name, y=perct_het_alt_in_snvs, color=gender), size=5) +
  xlab("Sample") +
  ylab("Heterozygosity Rate") +
  ggtitle("Heterozygosity Rate on the X Chromosome") +
  scale_colour_brewer(palette="Set1", name="Gender") +
  theme(axis.text.x=if(nrow(result) <= 20)
    {element_text(angle = 90, hjust = 1)} else {element_blank()},
    axis.line = element_line(colour = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank(),
    panel.background = element_blank(),
    axis.text=element_text(size=24),
    axis.ticks = element_blank(),
    axis.text.x = element_blank(),
    axis.title=element_text(size=28),
    title=element_text(size=32,face="bold"),
    legend.text=element_text(size=24)) 
```

#### Ti/Tv By Genomic Window
```{r message=FALSE, warning=FALSE, comment=NA}
titvWindowResults <- DisplayAndDispatchQuery("./sql/ti-tv-ratio.sql",
                                             project=project,
                                             replacements=c("#_WHERE_"="WHERE reference_name = 'chr1'",
                                                            "_WINDOW_SIZE_"="100000",
                                                            queryReplacements))
```

```{r titv-by-genomic-window-publication, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(titvWindowResults, aes(x=window_start, y=titv)) +
  geom_point() +
  stat_smooth() +
  scale_x_continuous(labels=comma) +
  xlab("Genomic Position") +
  ylab("Ti/Tv") +
  scale_x_continuous(expand = c(0, 0)) +
  ggtitle("Ti/Tv by 100,000 base pair windows\non Chromosome 1") +
  theme(axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        legend.position = "none",
        axis.text=element_text(size=20),
        axis.title=element_text(size=28),
        title=element_text(size=32,face="bold"))
```

#### IBS
```{r message=FALSE, warning=FALSE, comment=NA}
ibsDataFlowFilename = '/Users/gmcinnes/data/all-genomes-ibs-2.tsv'
ReadIBSFile <- function(ibsFilename, header=FALSE, rowNames=NULL) {
  ibsData <- read.table(file=ibsFilename, header=header,
                        row.names=rowNames, stringsAsFactors=FALSE)
  return (ibsData)
}
ibsDataflowData <- ReadIBSFile(ibsDataFlowFilename)

ColumnNames <- function(ibsData) { 
  if(3 == ncol(ibsData)) {
    colnames(ibsData) <- c("sample1", "sample2", "ibsScore")
  } else {
    colnames(ibsData) <- c("sample1", "sample2", "ibsScore", "similar", "observed")
  }
}
colnames(ibsDataflowData) <- ColumnNames(ibsDataflowData)

MakeIBSDataSymmetric <- function(ibsData) {
  ibsPairsMirrored <- data.frame(sample1=ibsData$sample2,
                                 sample2=ibsData$sample1,
                                 ibsScore=ibsData$ibsScore)
  ibsData <- rbind(ibsData[,1:3], ibsPairsMirrored)
}
ibsDataflowData <- MakeIBSDataSymmetric(ibsDataflowData)

ExcludeDiagonal <- function(ibsData) {
  ibsData <- filter(ibsData, ibsData$sample1 != ibsData$sample2)
  return (ibsData)
}
ibsDataflowDataSample <- ExcludeDiagonal(ibsDataflowData)


SampleIBSMatrix <- function(ibsData, sampleSize=50) {
  individuals <- unique(ibsData$sample1)
  sample <- sample(individuals, sampleSize)
  # In order to highlight related samples we make sure these are included in the subset
  related = c('LP6005051-DNA_D09',
              'LP6005051-DNA_E02', 
              'LP6005692-DNA_D05', 
              'LP6005243-DNA_E10',
              'LP6005144-DNA_A02',
              "LP6005051-DNA_D04",
              "LP6005243-DNA_H03",
              "LP6005144-DNA_D04",
              "LP6005692-DNA_E10",
              "LP6005692-DNA_G09")
  sample = c(sample, related)
  ibsData <- subset(ibsData, ibsData$sample1 %in% sample)
  ibsData <- subset(ibsData, ibsData$sample2 %in% sample)
  return (ibsData)
}
ibsDataflowDataSubset <- SampleIBSMatrix(ibsDataflowDataSample)

DrawHeatMap <- function(ibsData) {
  p <- ggplot(data=ibsData, aes(x=sample1, y=sample2)) +
    theme_bw() +
    theme(axis.ticks=element_blank(), 
          axis.text=element_blank(),
          axis.line = element_line(colour = "black"),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_blank(),
          panel.background = element_blank(),
          title=element_text(size=24,face="bold")) +
    geom_tile(aes(fill=ibsScore), colour="white") +
    scale_fill_gradient(low="white", high="steelblue", na.value="black",
                        guide=guide_colourbar(title= "IBS Score")) +
    labs(list(title="Identity By State (IBS) Heat Map",
              x="Sample", y="Sample")) 
  p 
}
DrawHeatMap <- function(ibsData) {
  p <- ggplot(data=ibsData, aes(x=sample1, y=sample2)) +
    theme_bw() +
    theme(axis.ticks=element_blank(), 
          axis.text=element_blank(),
          axis.line = element_line(colour = "black"),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_blank(),
          panel.background = element_blank(),
          axis.title=element_text(size=28),
          legend.text=element_text(size=24),
          title=element_text(size=32,face="bold")) +
    geom_tile(aes(fill=ibsScore), colour="white") +
    scale_fill_gradient(low="white", high="steelblue", na.value="black",
                        guide=guide_colourbar(title= "IBS Score")) +
    labs(list(title="Identity By State (IBS) Heat Map",
              x="Sample", y="Sample")) 
  p 
}
```

```{r ibs-publication, fig.align="center", fig.width=10, message=FALSE, comment=NA}
DrawHeatMap(ibsDataflowDataSubset)
```
