<!-- R Markdown Documentation, DO NOT EDIT THE PLAIN MARKDOWN VERSION OF THIS FILE -->
  
  <!-- Copyright 2015 Stanford University All rights reserved. -->
  
  <!-- Licensed under the Apache License, Version 2.0 (the "License"); -->
  <!-- you may not use this file except in compliance with the License. -->
  <!-- You may obtain a copy of the License at -->
  
  <!--     http://www.apache.org/licenses/LICENSE-2.0 -->
  
  <!-- Unless required by applicable law or agreed to in writing, software -->
  <!-- distributed under the License is distributed on an "AS IS" BASIS, -->
  <!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
  <!-- See the License for the specific language governing permissions and -->
  <!-- limitations under the License. -->
  
# Comparison of Google Genomics tools vs Academic Tools
  
  ```{r echo=FALSE, eval=FALSE}
######################[ CHANGE ME ]##################################
# This codelab assumes that the current working directory is where the Rmd file resides.
setwd("/Users/gmcinnes/GitHub/mvp_aaa_codelabs/qc")

# Set the Google Cloud Platform project id under which these queries will run.
project <- "gbsc-gcp-project-mvp"
#####################################################################
```

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE}
# Set up for BigQuery access.
source("./rHelpers/setup.R")
```


```{r}
tableReplacement <- list("_THE_TABLE_"="gbsc-gcp-project-mvp:va_aaa_pilot_data.5_genome_test_gvcfs_no_calls",
                          "_THE_EXPANDED_TABLE_"="gbsc-gcp-project-mvp:va_aaa_pilot_data.5_genome_test_vcfs_no_calls")
sampleData <- read.csv("./data/patient_info.csv")
sampleInfo <- select(sampleData, call_call_set_name=Catalog.ID, gender=Gender)

constraints <- list("#_WHERE_"="WHERE 
                    reference_name = 'chr17'
                    AND start BETWEEN 41196311
                    AND 41277499")

googleRepository = "https://raw.githubusercontent.com/deflaux/codelabs/qc-codelab/R/PlatinumGenomes-QC/sql/"
```
The purpose of this codelab is to compare BigQuery queries to standard command line tools for genomic analysis.  

* [Singleton Rate](#singleton-rate)

## Singleton Rate

For each genome, count the number of variants shared by no other member of the cohort.  Too many private calls for a particular individual may indicate a problem.

```{r message=FALSE, warning=FALSE, comment=NA}
query <- paste(googleRepository, "private-variants-brca1.sql", sep = "")
result <- DisplayAndDispatchQuery(query,
                                  project=project,
                                  replacements=tableReplacement)
```
Number of rows returned by this query: `r nrow(result)`.

Displaying the first few results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```

Compare to [output](./data/brca1/singletons.singletons) from vcftools, which has 85 some of which are for 0/0 genotypes from reference matching blocks (see the [vcftools command line](./data/brca1/singletons.e9636222) used to create this file).

```{r}
expectedResult <- read.table("./data/brca1/singletons.singletons", header=TRUE)
# Convert to zero-based coordinates
expectedResult <- mutate(expectedResult, POS = POS - 1)
# Clean colnames to match
colnames(expectedResult) <- gsub('\\.+', '_', colnames(expectedResult))
```

How many singletons do the two results have in common?
```{r}
nrow(inner_join(result, expectedResult))
```

Which singletons were only identified by BigQuery?
```{r results="asis"}
onlyBQ <- anti_join(result, expectedResult)
print(xtable(onlyBQ), type="html", include.rownames=F)
```

Which singletons were only identified by vcftools?
```{r results="asis"}
onlyVcftools <- anti_join(expectedResult, result)
print(xtable(onlyVcftools), type="html", include.rownames=F)
```

Retrieving the gVCF data for the singletons identified only by vcftools:
```{r message=FALSE, warning=FALSE, comment=NA}
having <- paste("start = ", onlyVcftools$POS,
                sep="", collapse=" OR ")
query <- paste(googleRepository, "examine-data.sql", sep="")
result <- DisplayAndDispatchQuery(query,
                                  project=project,
                                  replacements=c(tableReplacement,
                                                 "_HAVING_"=having))
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```

Most of these positions correspond to sites where 4 of the 5 genomes had a no-call, this is not neccessarily a singleton.

It appears that they correspond either to:
* a reference-matching block, so not actually a singleton and just perhaps violating an assumption in the vcftools code
* or a non-singleon variant, perhaps due to a problem in converting the gVCF data to all-positions VCF via gvcftools?


## Homozygosity Rate and Inbreeding Coefficient

For each genome, compare the expected and observed rates of homozygosity.

```{r message=FALSE, warning=FALSE, comment=NA}
query <- paste(googleRepository, "homozygous-variants.sql", sep="") 
result <- DisplayAndDispatchQuery("./sql/homozygous-variants.sql",
                                  project=project,
                                  replacements=c(tableReplacement, constraints))
```
Number of rows returned by this query: `r nrow(result)`.

Displaying the first few results:
  ```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```


```{r}
expectedResult <- read.table("./data/brca1/het_calls.het", header=TRUE)
# Clean colnames to match
colnames(expectedResult) <- gsub('\\.+$', '', colnames(expectedResult))
colnames(expectedResult) <- gsub('\\.+', '_', colnames(expectedResult))
```

```{r results="asis"}
n = names(result)
n[1] = "INDV"
names(result) = n
joinedResult <- inner_join(expectedResult, result, by=c("INDV"))
print(xtable(joinedResult[,order(colnames(joinedResult))]), type="html", include.rownames=F)
```

The logic in the query looks similar to vcftools [output_het method](http://sourceforge.net/p/vcftools/code/HEAD/tree/trunk/cpp/variant_file_output.cpp#l165) but there is clearly a difference.  TODO: investigate the difference further.


Check Hardy-Weinberg Equilibrium
-----------------------------------
```{r message=FALSE, warning=FALSE, comment=NA}
query <- paste(googleRepository, "hardy-weinberg-brca1.sql", sep="")
result <- DisplayAndDispatchQuery(query,
                                  project=project,
                                  replacements=tableReplacement)
```
Number of rows returned by this query: `r nrow(result)`.

Displaying the first few results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```

Compare to [brca1.hwe](./data/brca1/hardy.hwe) (see the [vcftools command line](./data/hwe/brca1.log) used to create this file).

```{r}
require(dplyr)
df <- read.table("./data/brca1/hardy.hwe", header=TRUE)
obsSplitCol <- "OBS.HOM1.HET.HOM2."
obsTemp <- read.table(text=as.character(df[, obsSplitCol]), sep = "/")
names(obsTemp) <- c("OBS_HOM1", "OBS_HET", "OBS_HOM2")
eSplitCol <- "E.HOM1.HET.HOM2."
eTemp <- read.table(text=as.character(df[, eSplitCol]), sep = "/")
names(eTemp) <- c("E_HOM1", "E_HET", "E_HOM2")
expectedResult <- cbind(cbind(df[setdiff(names(df), c(obsSplitCol,eSplitCol))], obsTemp), eTemp)
# Convert to zero-based coordinates
expectedResult <- mutate(expectedResult, POS = POS - 1)
```

How many results do the two results have in common?
```{r}
nrow(inner_join(result, expectedResult, by=c("CHR", "POS", "OBS_HOM1", "OBS_HET", "OBS_HOM2")))
```

Which results were only identified by BigQuery?
```{r results="asis"}
onlyBQ <- anti_join(result, expectedResult, , by=c("CHR", "POS", "OBS_HOM1", "OBS_HET", "OBS_HOM2"))
print(xtable(arrange(onlyBQ, CHR, POS)), type="html", include.rownames=F)
```

Note vcftools appears to skip variants with single allele genotypes:
```
zgrep 41242078 platinum_genomes_brca1_expanded_merged.vcf.gz 
chr17  41242078  .  G  A  143	LowGQX;TruthSensitivityTranche99.90to100.00;LowQD;SiteConflict	BLOCKAVG_min30p3a;MQ=57;MQ0=0;BaseQRankSum=0.781;Dels=0.3;FS=1.561;HRun=11;HaplotypeScore=77.7361;MQRankSum=0.093;QD=2.01;ReadPosRankSum=-2.871;SB=-45.67;VQSLOD=-1.8762;culprit=QD;set=FilteredInAll;DP=425;AF=0.5;AN=25;AC=1	GT:DP:GQX:MQ:AD:GQ:PL:VF	0/0:57:99:59:.:.:.:.	0:27:25:57:26:25.38:.:.	0/0:51:99:57:.:.:.:.	0/1:50:99:59:42,8:99:173,0,1238:0.16	0/0:46:99:59:.:.:.:.	0/0:44:99:60:.:.:.:.	.:46:.:59:40,6:.:.:.	0/0:40:85:59:.:.:.:.	0/0:40:85:59:.:.:.:.	0/0:63:99:58:.:.:.:.	0:42:2:58:37:1.58:.:.	0:33:0:57:29:0.03:.:.	.:44:.:58:31,12:.:.:.	0/0:44:90:58:.:.:.:.	0/0:40:87:58:.:.:.:.	.:44:.:57:39,5:.:.:.	0/0:55:99:59:.:.:.:.
```

Which results were only identified by vcftools?
```{r results="asis"}
onlyVcftools <- anti_join(expectedResult, result, , by=c("CHR", "POS", "OBS_HOM1", "OBS_HET", "OBS_HOM2"))
print(xtable(arrange(onlyVcftools, CHR, POS)), type="html", include.rownames=F)
```

Retrieving the gVCF data for the results identified only by vcftools:
```{r message=FALSE, warning=FALSE, comment=NA}
having <- paste("start <= ", onlyVcftools$POS,
                "AND",
                "end >= ", onlyVcftools$POS+1)
query <- paste(googleRepository, "examine-data.sql", sep="")
result <- DisplayAndDispatchQuery(query,
                                  project=project,
                                  replacements=c(tableReplacement,
                                                 "_HAVING_"=having))
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```

It appears that with BigQuery we are computing HWE for all the same variants as vcftools and the expected and Chi-Squared values are only slightly different.

See also: the [gVCF version of this query](./sql/hardy-weinberg-brca1.sql), which is close but only works for SNPs and needs a RIGHT OUTER JOIN to compute values for variants for which all the samples have the variant.

Ti/Tv by Alternate Allele Counts
-----------------------------------
https://github.com/deflaux/codelabs/blob/qc-codelab/R/PlatinumGenomes-QC/Variant-Level-QC.md#titv-by-alternate-allele-counts
```{r message=FALSE, warning=FALSE, comment=NA}
query <- paste(googleRepository, "ti-tv-by-alternate-allele-count.sql", sep="")
result <- DisplayAndDispatchQuery(query,
                                  project=project,
                                  replacements=c(tableReplacement, constraints))
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```
    
Compare to [output](./data/brca1/TsTv-by-count.TsTv.count) from vcftools.

```{r}
expectedResult <- read.table("./data/brca1/TsTv-by-count.TsTv.count", header=TRUE)
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(expectedResult), type="html", include.rownames=F)
```




Missingness sample level
-----------------------------------
https://github.com/deflaux/codelabs/blob/qc-codelab/R/PlatinumGenomes-QC/Sample-Level-QC.md#missingness-rate
```{r message=FALSE, warning=FALSE, comment=NA}
query <- paste(googleRepository, "sample-level-missingness.sql", sep="")
result <- DisplayAndDispatchQuery(query,
                                  project=project,
                                  replacements=c(tableReplacement, constraints))
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```
    
Compare to [output](./data/brca1/missingness.imiss) from vcftools.

```{r}
expectedResult <- read.table("./data/brca1/missingness.imiss", header=TRUE)
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(expectedResult), type="html", include.rownames=F)
```

Missingness variant level
-----------------------------------
https://github.com/deflaux/codelabs/blob/qc-codelab/R/PlatinumGenomes-QC/Variant-Level-QC.md#missingness-rate
```{r message=FALSE, warning=FALSE, comment=NA}
query <- paste(googleRepository, "variant-level-missingness.sql", sep="")
result <- DisplayAndDispatchQuery(query,
                                  project=project,
                                  replacements=c(tableReplacement, constraints))
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```
    
Compare to [output](./data/brca1/missing_sites.lmiss) from vcftools.

```{r}
expectedResult <- read.table("./data/brca1/missing_sites.lmiss", header=TRUE)
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(expectedResult)), type="html", include.rownames=F)
```


Sex Inference
-----------------------------------
https://github.com/deflaux/codelabs/blob/qc-codelab/R/PlatinumGenomes-QC/Sample-Level-QC.md#sex-inference
```{r message=FALSE, warning=FALSE, comment=NA}
query <- paste(googleRepository, "gender-check.sql", sep="")
result <- DisplayAndDispatchQuery(query,
                                  project=project,
                                  replacements=c(tableReplacement, constraints))
```

Number of rows returned by this query: `r nrow(result)`.

Displaying the first few results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```

Let's join this with the sample information:
```{r message=FALSE, warning=FALSE, comment=NA}
joinedResult <- inner_join(result, sampleInfo)
```

And visualize the results:
```{r gender, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(joinedResult) +
  geom_point(aes(x=call_call_set_name, y=perct_het_alt_in_snvs, color=gender)) +
  theme(axis.text.x=if(nrow(result) <= 20)
    {element_text(angle = 90, hjust = 1)} else {element_blank()}) +
  xlab("Sample") +
  ylab("Heterozygosity Rate ") +
  ggtitle("Heterozygosity Rate on the X Chromosome")

```

Ethnicity Inference
-----------------------------------
https://github.com/deflaux/codelabs/blob/qc-codelab/R/PlatinumGenomes-QC/Sample-Level-QC.md#ethnicity-inference

Genome Similarity
-----------------------------------
https://github.com/deflaux/codelabs/blob/qc-codelab/R/PlatinumGenomes-QC/Sample-Level-QC.md#genome-similarity

Ti/Tv by Depth
-----------------------------------

Heterozygous haplotype
-----------------------------------








